db/version_set.cc

```diff
@@ -5993,21 +6010,164 @@ Status VersionSet::ProcessManifestWrites(
         }
       }
 
+      // Adjusting manifest encoding for transient CFs
+      //
+      // Transient CF version edits need to be excluded from the manifest, but
+      // are retained in memory to build versions. However, if edits are part
+      // of an atomic group, each edit tracks remaining_entries. We need to
+      // adjust remaining_entries before writing the edits to manifest, so the
+      // missing transient CF version edits are not reported as manifest
+      // corruption. We then restore previous values for remaining_entries, as
+      // we have validations that check the remaining entries for edits in an
+      // atomic group, and these expect the transient CF edits to be there.
+      //
+      // NOTE: We could remove the validations and not restore the previous
+      // values, but we choose to do so for consistency with the in-memory
+      // Version.
+      //
+      // We skip this entirely if there are no transient CFs in
+      // atomic groups, based on the has_transient_cf_edits_in_atomic_group bool
+      // we set above. This avoids the extra pass for most RocksDB instances
+      // where transient CFs will not be used.
+      //
+      // For each edit in an atomic group, we need to know how many transient
+      // CF edits will be skipped AFTER it, so we can adjust its
+      // remaining_entries before writing to MANIFEST. We iterate backward
+      // through batch_edits, tracking the number of transient edits seen so
+      // far. When we encounter a non-transient edit, we know how many transient
+      // edits follow it.
+      std::vector<uint32_t> remaining_adjustments(batch_edits.size(), 0);
```

Comment:
Given that GetRemainingEntries() is only used in this function, it might be cleaner / efficient to adjust it in-place in each version edit. We can skip the edit from transient CF to make this validation pass: https://github.com/facebook/rocksdb/blob/4951494a27cf1c4c886297a1a4e1e5fdd2473ebe/db/version_set.cc#L5847.